File: 3106-length-of-the-longest-subsequence-that-sums-to-target/length-of-the-longest-subsequence-that-sums-to-target.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var lengthOfLongestSubsequence = function(nums, target) {
    const n = nums.length;
    const dp = new Array(n+1);
    for(let i=0; i<=n; i++) {
        dp[i] = new Array(target+1).fill(-1);
    }

    for(let i=0; i<=n; i++) {
        dp[i][0] = 0;
    }

    for(let i=1; i<=target; i++) {
        for(let j=1; j<=n; j++) {
            dp[j][i] = dp[j-1][i];
            if(i >= nums[j-1] && dp[j-1][i-nums[j-1]] != -1) dp[j][i] = Math.max(dp[j-1][i-nums[j-1]] + 1, dp[j][i]);
        }
    }

    return dp[n][target];
};

File: 3106-length-of-the-longest-subsequence-that-sums-to-target/length-of-the-longest-subsequence-that-sums-to-target.java
class Solution {
    int[][] dp;
    int solve(List<Integer> nums, int idx, int target, int curr) {
        if(curr > target) return -1000;
        if(idx == nums.size()) {
            if(curr == target) return 0;
            else return -1000;
        }
        if(dp[idx][curr] != -1) return dp[idx][curr];
        return dp[idx][curr] = Math.max(solve(nums, idx+1, target, curr), 1 + solve(nums, idx+1, target, curr + nums.get(idx)));
    }
    public int lengthOfLongestSubsequence(List<Integer> nums, int target) {
        dp = new int[nums.size()][target+1];
        for(int i=0; i<dp.length; i++) Arrays.fill(dp[i], -1);
        int ans = solve(nums, 0, target, 0);
        return ans < 0 ? -1 : ans;
    }
}